### PR TITLE
docs: revise the README for introducing Kaito tuning

### PR DESCRIPTION
**Pull request by Fei-Guo**
This change revises the README for katio tuning. It adds details for Pod structure and steps for basic troubleshooting. It also refactors the existing content with certain simplifications,

### CODE REVIEW

File: docs/tuning/README.md [hunk 1]
```diff
@@ -1,17 +1,15 @@
-# Kaito Tuning Workspace API
+# Kaito fine tuning
+This document presents how to use the Kaito `workspace` CRD for parameter-efficient fine-tuning (PEFT) of models, how a Kubernetes job is designed to automate the tunning workflow and several best practices for troubleshooting.
```
Comment (ishaansehgal99) [inline docs/tuning/README.md:2]:
```suggestion
This document presents how to use the Kaito `workspace` Custom Resource Definition (CRD) for parameter-efficient fine-tuning (PEFT) of models, how a Kubernetes job is designed to automate the tuning workflow, and several best practices for troubleshooting.
```

File: docs/tuning/README.md [hunk 2]
```diff
@@ -1,17 +1,15 @@
...
 
-This guide provides instructions on how to use the Kaito Tuning Workspace API for parameter-efficient fine-tuning (PEFT) of models. The API supports methods like LoRA and QLoRA and allows users to specify their own datasets and configuration settings.
+## Usage
+Kaito tuning APIs allow users to specify the supported tuning methods like LoRA or QLoRA, the input dataset and configuration settings, and the output destination for saving the tunning results. Currently, Kaito supports both URL and image as the type of tuning input source. It only supports image as the type of output destination. In the future, Kaito will additionally support the Kubernetes `v1.Volume` API for both the input source and the output destination.
```
Comment (ishaansehgal99) [inline docs/tuning/README.md:5]:
```suggestion
Kaito tuning APIs allow users to specify supported tuning methods like LoRA or QLoRA, the input dataset and configuration settings, and the output destination for saving the tuning results. Currently, Kaito supports both URL and image as the types of tuning input sources. It only supports image as the type of output destination. In the future, Kaito will additionally support the Kubernetes `v1.Volume` API for both the input source and the output destination.

```

File: docs/tuning/README.md [hunk 3]
```diff
@@ -21,98 +19,67 @@ resource:
...
 
-- [Default LoRA ConfigMap](../../charts/kaito/workspace/templates/lora-params.yaml)
+### Tuning configurations
+Kaito provides default tuning configurations for different tuning methods. They are managed by Kubernetes configmaps.
+- [default LoRA configmap](../../charts/kaito/workspace/templates/lora-params.yaml)
+- [default QLoRA configmap](../../charts/kaito/workspace/templates/qlora-params.yaml)
 
-- [QLoRA ConfigMap](../../charts/kaito/workspace/templates/qlora-params.yaml)
+## Tuning configmaps
+User can specify a customized configmap via the `Config` field of the `TuningSpec`. The customized configmap should be structured based on the default configmaps provided by Kaito. Please read the following section carefully when attempting to change the default parameters used by Kaito.
```
Comment (ishaansehgal99) [inline docs/tuning/README.md:46]:
```suggestion
User can specify a customized configmap via the `Config` field of the `TuningSpec`. The customized ConfigMap should be structured based on the default ConfigMaps provided by Kaito. Please read the following section carefully when attempting to change the default parameters used by Kaito.

```

File: docs/tuning/README.md [hunk 4]
```diff
@@ -122,22 +89,52 @@ The dataset should follow a specific format. For example:
...
 
 
-## Dataset Format Support
+Note: if you build a container image for the input dataset, please copy the dataset to the **`/data`** directory inside the container.
 
-The SFTTrainer supports popular dataset formats, allowing direct passage of the dataset without pre-processing. Supported formats include conversational and instruction formats. If your dataset is not in one of these formats, it will be passed directly to the SFTTrainer without any preprocessing. This may result in undefined behavior if the dataset does not align with the trainer's expected input structure.
+# Tuning Job
+Kaito uses the Kubernetes **batchv1.job** workload to manage the tuning Pod. When a tuning workspace custom resource is created, Kaito controller will create a job with the same name of the workspace in the same namespace. To streamline the tuning workflow, Kaito adds two containers in addition to the main container that runs the tuning process. The pod structure is illustrated in Figure 1.
```
Comment (ishaansehgal99) [inline docs/tuning/README.md:110]:
```suggestion
Kaito uses the Kubernetes **batchv1.job** workload to manage the tuning Pod. When a tuning workspace custom resource is created, the Kaito controller will create a job with the same name as the workspace in the same namespace. To streamline the tuning workflow, Kaito adds two containers in addition to the main container that runs the tuning process. The pod structure is illustrated in Figure 1.
```

File: docs/tuning/README.md [hunk 5]
```diff
@@ -122,22 +89,52 @@ The dataset should follow a specific format. For example:
...
+</div>
+Figure 1. Kaito tuning pod structure.
 
-To ensure proper function, you may need to preprocess the dataset to match one of the supported formats.
+An initcontainer `data-downloader` is introduced to download the training input dataset from the URLs specified in the tuning spec. In case an image is specified in the input, the `data-downloader` container simply uses the specified image as the container image. This initcontainer ensures the training data is available locally before the training process starts.
+
+To support automatically pushing the tuning results to a container registry, Kaito adds a sidecar container which has `docker` installed. This container runs a script which periodically checks the training progress. Once the training is done, indicated by a sentinel file created by the training process, the script will build a container image containing the training results and push the image to the specified container registry.
+
+The main container uses one of the supported model images. The image entry launches the [fine\_tuning.py](https://github.com/Azure/kaito/blob/main/presets/tuning/text-generation/fine_tuning.py) script.
```
Comment (ishaansehgal99) [inline docs/tuning/README.md:120]:
```suggestion
- InitContainer `data-downloader`: Downloads the training input dataset from the URLs specified in the tuning spec. If an image is specified in the input, the `data-downloader` container uses the specified image as the container image. This initContainer ensures the training data is available locally before the training process starts.

- Sidecar container: Supports automatically pushing the tuning results to a container registry. This container, with `docker` installed, runs a script to periodically check the training progress. Once the training is done, indicated by a sentinel file created by the training process, the script builds a container image containing the training results and pushes the image to the specified container registry.

- Main container: Uses one of the supported model images. The image entry launches the [fine\_tuning.py](https://github.com/Azure/kaito/blob/main/presets/tuning/text-generation/fine_tuning.py) script.
```

File: docs/tuning/README.md [hunk 6]
```diff
@@ -122,22 +89,52 @@ The dataset should follow a specific format. For example:
...
+
+All three containers use shared local volumes (by mounting same `EmptyDir` volumes). Hence file copies between containers are avoided.
```
Comment (ishaansehgal99) [inline docs/tuning/README.md:122]: All three containers use shared local volumes (by mounting the same `EmptyDir` volumes), hence file copies between containers are avoided.

Comment (ishaansehgal99) [inline docs/tuning/README.md:122]:
```suggestion
All three containers use shared local volumes (by mounting the same `EmptyDir` volumes), hence file copies between containers are avoided.
```

File: docs/tuning/README.md [hunk 7]
```diff
@@ -122,22 +89,52 @@ The dataset should follow a specific format. For example:
...
+
+# Troubleshooting
+
+### Job pod failures
+When the tuning job reaches the failed state, at least one of the above three containers hit errors. Users can check the logs of these containers using the `kubectl logs PODNAME -n NAMESPACE -c CONTAINERNAME` command.
```
Comment (ishaansehgal99) [inline docs/tuning/README.md:127]:
```suggestion
When the tuning job reaches the failed state, at least one of the above three containers has encountered errors. Users can check the logs of these containers using the `kubectl logs PODNAME -n NAMESPACE -c CONTAINERNAME` command.
```

File: docs/tuning/README.md [hunk 8]
```diff
@@ -122,22 +89,52 @@ The dataset should follow a specific format. For example:
...
+
+For the inticontainer and the sidecar container, possible errors include invalid input/out URLs or invalid image pull secretes. Users can fix the problems by updating the workspace custom resource with corrections. The Kaito controller will create a new job using the updated spec.
+
+For the main container, errors may occur when CUDA reports out of gpu memory. In this case, users should reduce the batch size (the default is 1) if it has been customized to a value larger than 1. If the batch size is already 1, the workspace must be recreated using a different gpu sku with bigger gpu memory. Note that Kaito has optimized the training memory usage by dropping the preallocated memory cache. Our internal tests show that the performance impact due to this change is negligible.
+
+### Time for job completion
+The training job can take a long time depending on the size of the input dataset and training pipeline configurations. The total training time is largely determined by the total number of training steps that need to be performed. The total number of steps can be calculated as:
```
Comment (ishaansehgal99) [inline docs/tuning/README.md:134]:
```suggestion
The training job can take a long time depending on the size of the input dataset and training pipeline configurations. The total training time is largely determined by the total number of training steps, calculated as:
```

File: docs/tuning/README.md [hunk 9]
```diff
@@ -122,22 +89,52 @@ The dataset should follow a specific format. For example:
...
+The main container uses one of the supported model images. The image entry launches the [fine\_tuning.py](https://github.com/Azure/kaito/blob/main/presets/tuning/text-generation/fine_tuning.py) script.
+
+All three containers use shared local volumes (by mounting same `EmptyDir` volumes). Hence file copies between containers are avoided.
+
+# Troubleshooting
+
+### Job pod failures
+When the tuning job reaches the failed state, at least one of the above three containers hit errors. Users can check the logs of these containers using the `kubectl logs PODNAME -n NAMESPACE -c CONTAINERNAME` command.
+
+For the inticontainer and the sidecar container, possible errors include invalid input/out URLs or invalid image pull secretes. Users can fix the problems by updating the workspace custom resource with corrections. The Kaito controller will create a new job using the updated spec.
```
Comment (ishaansehgal99) [inline docs/tuning/README.md:129]:
```suggestion
- InitContainer and Sidecar Container: Possible errors include invalid input/output URLs or invalid image pull secrets. Users can fix these problems by updating the workspace custom resource with corrections. The Kaito controller will create a new job using the updated spec.
```

File: docs/tuning/README.md [hunk 10]
```diff
@@ -122,22 +89,52 @@ The dataset should follow a specific format. For example:
...
+
+For the main container, errors may occur when CUDA reports out of gpu memory. In this case, users should reduce the batch size (the default is 1) if it has been customized to a value larger than 1. If the batch size is already 1, the workspace must be recreated using a different gpu sku with bigger gpu memory. Note that Kaito has optimized the training memory usage by dropping the preallocated memory cache. Our internal tests show that the performance impact due to this change is negligible.
```
Comment (ishaansehgal99) [inline docs/tuning/README.md:131]:
```suggestion
- Main Container: Errors may occur when CUDA reports out of GPU memory. Users should reduce the batch size (the default is 1) if it has been customized to a value larger than 1. If the batch size is already 1, the workspace must be recreated using a different GPU SKU with larger GPU memory. Note that Kaito has optimized the training memory usage by dropping the preallocated memory cache. Our internal tests show that the performance impact due to this change is negligible.
```

**Review by ishaansehgal99 — APPROVED** [review]

File: docs/tuning/README.md [hunk 11]
```diff
@@ -122,22 +90,52 @@ The dataset should follow a specific format. For example:
...
+# Tuning Job
+Kaito uses the Kubernetes **batchv1.job** workload to manage the tuning Pod. When a tuning workspace custom resource is created, the Kaito controller will create a job with the same name as the workspace in the same namespace. To streamline the tuning workflow, Kaito adds two containers in addition to the main container that runs the tuning process. The pod structure is illustrated in Figure 1.
+<div align="left">
+  <img src="../img/kaito-fine-tuning.png" width=40% title="Kaito fine tuning" alt="Kaito fine tuning">
+</div>
+Figure 1. Kaito tuning pod structure.
 
-To ensure proper function, you may need to preprocess the dataset to match one of the supported formats.
+- InitContainer `data-downloader`: Downloads the training input dataset from the URLs specified in the tuning spec. If an image is specified in the input, the `data-downloader` container uses the specified image as the container image. This initContainer ensures the training data is available locally before the training process starts.
```
Comment (Fei-Guo) [inline docs/tuning/README.md:117]:
```suggestion
- Initcontainer `data-downloader`: It downloads the training input dataset from the URLs specified in the tuning spec if needed. If an image is specified in the input, the `data-downloader` container uses the specified image as the container image. This initcontainer ensures the training data is available locally before the training process starts.
```

File: docs/tuning/README.md [hunk 12]
```diff
@@ -122,22 +90,52 @@ The dataset should follow a specific format. For example:
...
+<div align="left">
+  <img src="../img/kaito-fine-tuning.png" width=40% title="Kaito fine tuning" alt="Kaito fine tuning">
+</div>
+Figure 1. Kaito tuning pod structure.
 
-To ensure proper function, you may need to preprocess the dataset to match one of the supported formats.
+- Initcontainer `data-downloader`: It downloads the training input dataset from the URLs specified in the tuning spec if needed. If an image is specified in the input, the `data-downloader` container uses the specified image as the container image. This initcontainer ensures the training data is available locally before the training process starts.
+
+- Sidecar container: Supports automatically pushing the tuning results to a container registry. This container, with `docker` installed, runs a script to periodically check the training progress. Once the training is done, indicated by a sentinel file created by the training process, the script builds a container image containing the training results and pushes the image to the specified container registry.
```
Comment (Fei-Guo) [inline docs/tuning/README.md:119]:
```suggestion
- Sidecar container: It is introduced to support automatically pushing the tuning results to a container registry. This container, with `docker` installed, runs a script to periodically check the training progress. Once the training is done, indicated by a sentinel file created by the training process, the script builds a container image containing the training results and pushes the image to the specified container registry.
```

File: docs/tuning/README.md [hunk 13]
```diff
@@ -122,22 +90,52 @@ The dataset should follow a specific format. For example:
...
+</div>
+Figure 1. Kaito tuning pod structure.
 
-To ensure proper function, you may need to preprocess the dataset to match one of the supported formats.
+- Initcontainer `data-downloader`: It downloads the training input dataset from the URLs specified in the tuning spec if needed. If an image is specified in the input, the `data-downloader` container uses the specified image as the container image. This initcontainer ensures the training data is available locally before the training process starts.
+
+- Sidecar container: It is introduced to support automatically pushing the tuning results to a container registry. This container, with `docker` installed, runs a script to periodically check the training progress. Once the training is done, indicated by a sentinel file created by the training process, the script builds a container image containing the training results and pushes the image to the specified container registry.
+
+- Main container: Uses one of the supported model images. The image entry launches the [fine\_tuning.py](https://github.com/Azure/kaito/blob/main/presets/tuning/text-generation/fine_tuning.py) script.
```
Comment (Fei-Guo) [inline docs/tuning/README.md:121]:
```suggestion
- Main container: It uses one of the supported model images. The image entry launches the [fine\_tuning.py](https://github.com/Azure/kaito/blob/main/presets/tuning/text-generation/fine_tuning.py) script.
```

File: docs/tuning/README.md [hunk 14]
```diff
@@ -122,22 +90,52 @@ The dataset should follow a specific format. For example:
...
+- Main container: It uses one of the supported model images. The image entry launches the [fine\_tuning.py](https://github.com/Azure/kaito/blob/main/presets/tuning/text-generation/fine_tuning.py) script.
+
+All three containers use shared local volumes (by mounting the same `EmptyDir` volumes), hence file copies between containers are avoided.
+
+# Troubleshooting
+
+### Job pod failures
+When the tuning job reaches the failed state, at least one of the above three containers has encountered errors. Users can check the logs of these containers using the `kubectl logs PODNAME -n NAMESPACE -c CONTAINERNAME` command.
+
+- InitContainer and Sidecar Container: Possible errors include invalid input/output URLs or invalid image pull secrets. Users can fix these problems by updating the workspace custom resource with corrections. The Kaito controller will create a new job using the updated spec.
```
Comment (Fei-Guo) [inline docs/tuning/README.md:130]:
```suggestion
For the initcontainer and sidecar container, possible errors include invalid input/output URLs or invalid image pull secrets. Users can fix these problems by updating the workspace custom resource with corrections. The Kaito controller will create a new job using the updated spec.
```

File: docs/tuning/README.md [hunk 15]
```diff
@@ -1,17 +1,16 @@
-# Kaito Tuning Workspace API
+# Kaito fine tuning
+This document presents how to use the Kaito `workspace` Custom Resource Definition (CRD) for parameter-efficient fine-tuning (PEFT) of models, how a Kubernetes job is designed to automate the tuning workflow, and several best practices for troubleshooting.
 
-This guide provides instructions on how to use the Kaito Tuning Workspace API for parameter-efficient fine-tuning (PEFT) of models. The API supports methods like LoRA and QLoRA and allows users to specify their own datasets and configuration settings.
+## Usage
+Kaito tuning APIs allow users to specify supported tuning methods like LoRA or QLoRA, the input dataset and configuration settings, and the output destination for saving the tuning results. Currently, Kaito supports both URL and image as the types of tuning input sources. It only supports image as the type of output destination. In the future, Kaito will additionally support the Kubernetes `v1.Volume` API for both the input source and the output destination.
```
Comment (Fei-Guo) [inline docs/tuning/README.md:5]:
```suggestion
Kaito tuning APIs allow users to specify supported tuning methods like [LoRA or QLoRA](https://huggingface.co/docs/peft/main/en/conceptual_guides/lora), the input dataset and configuration settings, and the output destination for saving the tuning results. Currently, Kaito supports both URL and image as the types of tuning input sources. It only supports image as the type of output destination. In the future, Kaito will additionally support the Kubernetes `v1.Volume` API for both the input source and the output destination.
```

**Pull request merged**